zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/AuthorizationCheckBehavior.java

```diff
@@ -0,0 +1,151 @@
+   * the authorizations of the command.
+   *
+   * <p>This method does not take a Map of resource identifiers to check for. The user is considered
+   * authorized if it is the default user, or if it has a wildcard permission for the provided
+   * resource type and permission type. If you want to check for specific resource identifiers, use
+   * {@link #isAuthorized(TypedRecord, AuthorizationResourceType, PermissionType, Map)}
+   *
+   * @param command the command to check authorization for
+   * @param resourceType the type of resource to check authorization for
+   * @param permissionType the type of permission to check authorization for (CRUD)
+   * @return true if the user is authorized, false otherwise
+   */
+  public <T extends UnifiedRecordValue> boolean isAuthorized(
+      final TypedRecord<T> command,
+      final AuthorizationResourceType resourceType,
+      final PermissionType permissionType) {
+    return isAuthorized(command, resourceType, permissionType, Map.of());
+  }
+
+  /**
+   * Checks if a user is Authorized to perform an action on a resource. The user key is taken from *
+   * the authorizations of the command.
+   *
+   * <p>The caller of this method should provide a Map of resource identifiers. Examples of this
+   * are:
+   *
+   * <ul>
+   *   <li>Key: bpmnProcessId, Value: myProcess
+   *   <li>Key: processInstanceKey, Value: 1234567890
+   * </ul>
+   *
+   * @param command the command to check authorization for
+   * @param resourceType the type of resource to check authorization for
+   * @param permissionType the type of permission to check authorization for (CRUD)
+   * @param requiredResourceIdentifiers the resource identifiers to check for
+   * @return true if the user is authorized, false otherwise
+   */
+  public <T extends UnifiedRecordValue> boolean isAuthorized(
+      final TypedRecord<T> command,
+      final AuthorizationResourceType resourceType,
+      final PermissionType permissionType,
+      final Map<String, String> requiredResourceIdentifiers) {
+    final var userKey = (Long) command.getAuthorizations().get(Authorization.AUTHORIZED_USER_KEY);
+    if (userKey == null) {
+      return false;
+    }
+    return isAuthorized(userKey, resourceType, permissionType, requiredResourceIdentifiers);
+  }
+
+  private boolean isAuthorized(
+      final long userKey,
+      final AuthorizationResourceType resourceType,
+      final PermissionType permissionType,
+      final Map<String, String> requiredResourceIdentifiers) {
+
+    // If authorization is disabled, everyone is authorized
+    if (!engineConfig.isEnableAuthorization()) {
+      return true;
+    }
+
+    final var userOptional = userState.getUser(userKey);
+    if (userOptional.isEmpty()) {
+      return false;
+    }
+    final var user = userOptional.get();
+
+    // The default user has all permissions
+    if (user.getUserType().equals(UserType.DEFAULT)) {
+      return true;
+    }
```

Comment:
We could, but in the end we want customers to be able to create admin users as well. It seems to me that assigning a set of default permissions is very maintenance heavy, since any new permissions would need to be added here as well. Additionally it may require state migrations to get these permissions itno existing admin users.

I'd be curious to hear what struggles you've encountered with this in C7. For now, let's not let this discussion block this PR from merging so we can keep progressing.

Reply:
> We could, but in the end we want customers to be able to create admin users as well
Which you still could if you define a set of permissions internally that represent admin users, or even create a role for that with a set of permission and add admin users to that role.

> It seems to me that assigning a set of default permissions is very maintenance heavy, since any new permissions would need to be added here as well. Additionally it may require state migrations to get these permissions into existing admin/default users.

That is true, yes. It's the biggest downside of this approach, especially when we're in a developing phase where we add new permission types over a longer period of time.

> I'd be curious to hear what struggles you've encountered with this in C7

The biggest issue is that you must consider this specific role additionally, in command processing, Query API, and web apps. Whenever you touch on permission checks, you need to bake in the admin role on top. If it's just a user with a set of permissions, you don't. It makes all permission handling logic more complex by default. We have forgotten that role here and there sometimes when building new features and had to patch the admin role back in later. Contributing a feature to the platform that touches on permissions in any way is definitely more complex for anyone, especially externals. The admin role becomes "hidden knowledge" quite easily that way.

> For now, let's not let this discussion block this PR from merging so we can keep progressing.

I agree 👍